### PR TITLE
uucore: fix order of group IDs return from entries::get_groups()

### PR DESCRIPTION
**Pull request by jhscheer**
As discussed here: #2361
the group IDs returned for GNU's 'group' and GNU's 'id --groups'
starts with the effective group ID.
This implements a wrapper for `entris::get_groups()` which mimics
GNU's behaviour.

* add tests for `id`
* add tests for `groups`
* fix `id --groups --real` to no longer ignore `--real`

### CODE REVIEW

File: src/uucore/src/lib/features/entries.rs [hunk 1]
```diff
@@ -72,6 +72,31 @@ pub fn get_groups() -> IOResult<Vec<gid_t>> {
...
+    //
+    // If `arg_id` is None (default), `get_groups_gnu` moves the effective
+    // group id (egid) to the first entry in the returned Vector.
+    // If `arg_id` is Some(x), `get_groups_gnu` moves the id with value `x`
+    // to the first entry in the returned Vector. This might be necessary
+    // for `id --groups --real` if `gid` and `egid` are not equal.
+
+    let mut groups = get_groups()?;
+    if groups.len() > 1 {
+        let egid = arg_id.unwrap_or_else(|| crate::features::process::getegid());
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/entries.rs:88]: Please fix the clippy warning :)

File: src/uucore/src/lib/features/entries.rs [hunk 2]
```diff
@@ -72,6 +72,31 @@ pub fn get_groups() -> IOResult<Vec<gid_t>> {
...
+        if *groups.first().unwrap() != egid {
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/entries.rs:89]: Nice optimizations!

File: src/uucore/src/lib/features/entries.rs [hunk 3]
```diff
@@ -72,6 +72,31 @@ pub fn get_groups() -> IOResult<Vec<gid_t>> {
...
+            if let Some(index) = groups.iter().position(|&x| x == egid) {
+                groups.remove(index);
+                groups.insert(0, egid);
+            }
+        }
+    }
+    Ok(groups)
+}
+
+#[derive(Copy, Clone)]
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/entries.rs:99]: What was your reason for adding this?

File: src/uucore/src/lib/features/entries.rs [hunk 4]
```diff
@@ -72,6 +72,31 @@ pub fn get_groups() -> IOResult<Vec<gid_t>> {
...
+pub fn get_groups_gnu(arg_id: Option<u32>) -> IOResult<Vec<gid_t>> {
+    // The list of group IDs return from GNU's `groups` and GNU's `id --groups`
+    // starts with the effective group ID (egid).
+    // This is a wrapper for `get_groups()` to mimic this behavior.
+    //
+    // If `arg_id` is None (default), `get_groups_gnu` moves the effective
+    // group id (egid) to the first entry in the returned Vector.
+    // If `arg_id` is Some(x), `get_groups_gnu` moves the id with value `x`
+    // to the first entry in the returned Vector. This might be necessary
+    // for `id --groups --real` if `gid` and `egid` are not equal.
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/entries.rs:84]: Great explanation! It would be great to turn it into a rustdoc comment for this function!

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
Nice! Were you able to confirm that this solves the issue? Code-wise, I just have some small nits. Also there's this section from the libc manpages:
> It is unspecified whether the effective group ID of the calling process is included in the returned list.  (Thus, an application should also call getegid(2) and add or remove the resulting value.)

So I think the `egid` should always be added, even when it's not in the list? Would that be the desired behaviour?

**Comment by jhscheer** [comment]
> Nice! Were you able to confirm that this solves the issue? 

On my local hosts, yes. So far some CI tests don't look happy, I have to investigate that further.

> Also there's this section from the libc manpages:
> > It is unspecified whether the effective group ID of the calling process is included in the returned list.  (Thus, an application should also call getegid(2) and add or remove the resulting value.)
> 
> So I think the `egid` should always be added, even when it's not in the list? Would that be the desired behaviour?

I also came across that note in the man pages, but wasn't too sure what to make of it.

File: src/uucore/src/lib/features/entries.rs (hunk 3 above)
Comment (jhscheer) [reply src/uucore/src/lib/features/entries.rs:99]: Without this, there's a "use of moved value: `possible_pw` value used here after move" in L205 (and L199).

**Comment by tertsdiepraam** [comment]
Alright, then we should probably just keep it like this

**Comment by tertsdiepraam** [comment]
Locally, I also get this:
```
---- test_groups::test_groups stdout ----
current_directory_resolved: 
run: /home/terts/Programming/coreutils/target/debug/coreutils groups
current_directory_resolved: 
run: groups
thread 'test_groups::test_groups' panicked at 'assertion failed: `(left == right)`

Diff < left / right > :
<"terts sys network power lp wheel\n"
>"sys network power lp wheel terts\n"

', tests/common/util.rs:206:9
```
Which might be because my `groups` command is not coreutils but shadow-utils, apparently. Maybe something similar is going on in the CI?

**Comment by tertsdiepraam** [comment]
Maybe this helps. The OpenBSD implementation does this to construct the list of groups: https://github.com/openbsd/src/blob/fdf1c64914f5ad56fb382b8fd0e46398bc71a4cf/usr.bin/id/id.c#L305-L311

**Comment by jhscheer** [comment]
> Maybe this helps. The OpenBSD implementation does this to construct the list of groups: https://github.com/openbsd/src/blob/fdf1c64914f5ad56fb382b8fd0e46398bc71a4cf/usr.bin/id/id.c#L305-L311

Thanks. That's BSD licensed right? I'll take a look if I can't figure this out otherwise.

**Comment by jhscheer** [comment]
> > Also there's this section from the libc manpages:
> > > It is unspecified whether the effective group ID of the calling process is included in the returned list.  (Thus, an application should also call getegid(2) and add or remove the resulting value.)
> > 
> > 
> > So I think the `egid` should always be added, even when it's not in the list? Would that be the desired behaviour?
> 
> I also came across that note in the man pages, but wasn't too sure what to make of it.


I implemented this:
> As implied by the definition of supplementary groups, the
> effective group ID may appear in the array returned by
> getgroups() or it may be returned only by getegid().  Duplication
> may exist, but the application needs to call getegid() to be sure
> of getting all of the information. Various implementation
> variations and administrative sequences cause the set of groups
> appearing in the result of getgroups() to vary in order and as to
> whether the effective group ID is included, even when the set of
> groups is the same (in the mathematical sense of ``set''). (The
> history of a process and its parents could affect the details of
> the result.)
(From: https://www.man7.org/linux/man-pages/man3/getgroups.3p.html)

Everything looks good now :)

File: src/uucore/src/lib/features/entries.rs [hunk 5]
```diff
@@ -72,6 +72,41 @@ pub fn get_groups() -> IOResult<Vec<gid_t>> {
     }
 }
 
+/// The list of group IDs returned from GNU's `groups` and GNU's `id --groups`
+/// starts with the effective group ID (egid).
+/// This is a wrapper for `get_groups()` to mimic this behavior.
+///
+/// If `arg_id` is None (default), `get_groups_gnu` moves the effective
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/entries.rs:79]:
```suggestion
/// If `arg_id` is `None` (default), `get_groups_gnu` moves the effective
```
Might as well get really nitpicky :)

File: src/uucore/src/lib/features/entries.rs [hunk 6]
```diff
@@ -72,6 +72,41 @@ pub fn get_groups() -> IOResult<Vec<gid_t>> {
...
+/// group id (egid) to the first entry in the returned Vector.
+/// If `arg_id` is Some(x), `get_groups_gnu` moves the id with value `x`
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/entries.rs:81]:
```suggestion
/// If `arg_id` is `Some(x)`, `get_groups_gnu` moves the id with value `x`
```

File: src/uucore/src/lib/features/entries.rs [hunk 7]
```diff
@@ -268,3 +303,23 @@ pub fn usr2uid(name: &str) -> IOResult<uid_t> {
...
     Group::locate(name).map(|p| p.gid())
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn do_vecs_match<T: PartialEq>(a: &Vec<T>, b: &Vec<T>) -> bool {
+        let matching = a.iter().zip(b.iter()).filter(|&(a, b)| a == b).count();
+        matching == a.len() && matching == b.len()
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/features/entries.rs:313]: Why is this necessary? `Vec` implements `PartialEq`, right?

File: tests/by-util/test_id.rs [hunk 8]
```diff
@@ -104,19 +104,58 @@ fn test_id_group() {
...
+    //         .cmd_keepenv("strace")
+    //         .env("LANGUAGE", "C")
+    //         .args(args)
+    //         .run();
+    //     println!(
+    //         "stdout:\n{}\nstderr:\n{}",
+    //         strace.stdout_str(),
+    //         strace.stderr_str()
+    //     );
+    // }
```
Comment (tertsdiepraam) [inline tests/by-util/test_id.rs:131]:
```suggestion
```

File: tests/by-util/test_id.rs [hunk 9]
```diff
@@ -167,3 +206,32 @@ fn test_id_password_style() {
 
     assert!(result.stdout_str().starts_with(&username));
 }
+
+#[cfg(any(target_vendor = "apple", target_os = "linux"))]
```
Comment (tertsdiepraam) [inline tests/by-util/test_id.rs:210]: Is this not possible for unix systems other than Linux and Mac? Same question for the other function.

File: tests/by-util/test_groups.rs [hunk 10]
```diff
@@ -1,41 +1,53 @@
...
-    let result = TestScenario::new("id").ucmd_keepenv().arg("-un").run();
-    println!("result.stdout = {}", result.stdout_str());
-    println!("result.stderr = {}", result.stderr_str());
-    let s1 = String::from(result.stdout_str().trim());
-    if is_ci() && s1.parse::<f64>().is_ok() {
-        // In the CI, some server are failing to return id -un.
-        // So, if we are getting a uid, just skip this test
-        // As seems to be a configuration issue, ignoring it
+#[cfg(any(target_os = "linux"))]
+#[ignore = "fixme: 'groups USERNAME' needs more debugging"]
```
Comment (tertsdiepraam) [inline tests/by-util/test_groups.rs:17]: Should this still be ignored or it is left over from a previous commit?

Comment (tertsdiepraam) [reply tests/by-util/test_groups.rs:17]: Also, is this not available on other `unix` systems?

File: tests/by-util/test_groups.rs [hunk 11]
```diff
@@ -1,41 +1,53 @@
...
+fn test_groups_username() {
+    let scene = TestScenario::new(util_name!());
+    let whoami_result = scene.cmd("whoami").run();
```
Comment (tertsdiepraam) [inline tests/by-util/test_groups.rs:20]: Any possibility to get this without the `whoami` utility? Via libc? Although I guess that could be considered a little bit circular, as we're also using `libc` to get it in the first place. Maybe it would also work on the CI then?

**Review by tertsdiepraam — COMMENTED** [review]
Nice! Just some small nits and questions

File: src/uucore/src/lib/features/entries.rs (hunk 7 above)
Comment (tertsdiepraam) [reply src/uucore/src/lib/features/entries.rs:313]: I tested locally and
```rust
assert_eq!(get_groups_gnu(Some(last)).unwrap(), groups);
```
should work

File: tests/by-util/test_id.rs (hunk 9 above)
Comment (jhscheer) [reply tests/by-util/test_id.rs:210]: I find it extremely convenient to directly test against GNU coreutils 8.32.
`cmd_keepenv(util_name)` runs whatever util is in `$PATH` on the system the test runs on.
GNU coreutils is installed on the docker ubuntu images that github actions is using and I patched github actions to also install it for the macOS tests.
On every other system I can't guarantee that the util version is GNU. (If it's not GNU the tests may differ, you experienced this before when you said you had shadow coreutils installed).

File: tests/by-util/test_groups.rs (hunk 10 above)
Comment (jhscheer) [reply tests/by-util/test_groups.rs:17]: Should be ignored for now. There's a configuration bug on the docker ubuntu images which makes it difficult to test `id --name`, `groups <username>`, etc. I mentioned this e.g. here:
https://github.com/uutils/coreutils/blob/2f58f92b06ea8d5e625869e8920f9b2c00bc923a/tests/by-util/test_id.rs#L7-L10
Fixing this will require another PR.

File: tests/by-util/test_groups.rs (hunk 11 above)
Comment (jhscheer) [reply tests/by-util/test_groups.rs:20]: see my other comment

File: src/uucore/src/lib/features/entries.rs (hunk 7 above)
Comment (jhscheer) [reply src/uucore/src/lib/features/entries.rs:313]: thanks!

File: tests/by-util/test_id.rs (hunk 9 above)
Comment (tertsdiepraam) [reply tests/by-util/test_id.rs:175]: I agree that it's very convenient. However, I see a couple of issues with this:
- Other systems don't get tested as much and we might overlook subtle differences. The shadow-utils groups command also comes with Manjaro (and probably other Arch-based distros) by default, so the tests don't work there.
- It becomes circular when people run the tests on their machine on which they have uutils installed instead of GNU.


I understand that sometimes it's our best option, but I'd prefer to use it as little as possible.

**Review by tertsdiepraam — APPROVED** [review]

**Comment by tertsdiepraam** [comment]
Actually, I have one more request (sorry). Would you mind changing that call to GNU `groups` into a call to `id -Gn`? It should be identical[1] but works on Arch-based systems. Or is there a specific reason it needs to be `util_name!`? In that case, we could keep it like this.

[1]: https://www.gnu.org/software/coreutils/manual/html_node/groups-invocation.html#groups-invocation

File: tests/by-util/test_id.rs (hunk 9 above)
Comment (jhscheer) [reply tests/by-util/test_id.rs:175]: All true.

However, there's not really an alternative for tests with dynamically expected results (except quirky work-arounds). Best example is `id --groups`. Before I changed the tests, it was only tested if the stdout could be converted into numbers (fun fact, the parsing target was **f**64).

Imho. it's not a priority to have tests that support every linux distro, but it's of course important to document (`CONTRIBUTING.md`?) that some tests rely on GNU coreutils in the `$PATH`.
 
It's probably a good idea to implement a "feature" to tag the tests that rely on having GNU coreutils in the `$PATH`. So that it's easily possible to switch these tests off when running all tests in an environment without this prerequisite.

Comment (tertsdiepraam) [reply tests/by-util/test_id.rs:175]: I like the feature flag idea!

**Comment by jhscheer** [comment]
> Actually, I have one more request (sorry). Would you mind changing that call to GNU `groups` into a call to `id -Gn`? It should be identical [1] but works on Arch-based systems, where `groups` is not GNU. Or is there a specific reason it needs to be `util_name!`? In that case, we could keep it like this.
> 
> [1]: https://www.gnu.org/software/coreutils/manual/html_node/groups-invocation.html#groups-invocation

I think if you want to tinker with `test_groups_username`, it would be better not to do it in this PR.
Should I remove this test and just keep the `TODO` comments?

**Comment by tertsdiepraam** [comment]
Right, yeah I see that's non-trivial. Since it's just my machine I'll make a separate PR. You don't have to change this one
